contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+   the pod object with device selection info. eg:
+   `scheduler.alpha.kubernetes.io/resClass_test-res-class_nvidia-tesla-gpu=4`
+   where `scheduler.alpha.kubernetes.io/resClass` is the common prefix for all the
+   device annotations, `tes-res-class` is resource class name,
+   `nvidia-tesla-gpu` is the selected device name and `4` is the quantity requested.
+
+4. Kubelet reads the device request from pod annotation and calls `Allocate` on
+   the matching Device Plugins.
+5. The user deletes the pod or the pod terminates
+6. Kubelet reads pod object annotation for devices consumed and calls `Deallocate`
+   on the matching Device Plugins
+
+In addition to node selection, the scheduler is also responsible for selecting a
+device that matches the resource class requested by the user.
+
+### Reason for preferring device selection at the scheduler and not at the kubelet
+Kubelet does not maintain any cache. Therefore to know the availability of a
+device, which is requested by the new incoming pod, kubelet calculates how many
+devices are consumed by all already admitted pods, by iterating over all the admitted
+pods running on the node. This is done while running predicates for each new
+incoming pod at kubelet. Even if we assume that scheduler cache and consumption
+state that is created at runtime for each pod, are exactly same, current api
+interfaces does not allow to pass selected device to container manager (where
+actually device plugin will be invoked from). This problem occurs because
+requested resource classes are translated into devices internally through code
+and user does not mention device in pod object. While other resource requests
+can be determined from the pod object directly.
+To summarize, device selection at the kubelet can be done in one of the following
+two ways:
+* Select device at pod admission while applying predicates and change all api
+  interfaces that are required to pass selected device to container manager.
+* Create resource consumption state again at container manager and select device.
+
+None of the above approach seems cleaner than doing device selection at scheduler,
+which helps to retain cleaner api interfaces between packages.
+
+## Scheduler Changes
+Scheduler already listens and maintains state in the cache for any changes in
+node or pod objects. We will enhance the logic:
+1. To listen and maintain the state in cache for user created *Resource Class* objects.
+2. To look for device related details in node objects and maintain accounting for
+   devices as well.
+
+From the events perspective, handling for the following events will be added/updated:
+
+### Resource Class Creation
+1. Initialize and add resource class info into local cache
+2. Iterate over all existing nodes in cache to figure out if there are devices
+   on these nodes which are selectable by resource class. If found, update the
+   resource class availability status in local cache.
```

Comment:
@vikaschoudhary16 I'm very interested in the details about this? As commented above, how does Resource Class know which devices is selectable?

Reply:
Does above reply answer this question?

Reply:
Yes, thanks!